### PR TITLE
fix race condition and nil pointer dereference in holepunch

### LINKED ISSUE #1448: holepunch: s.holePuncher is nil on close
## Sys info
go-libp2p@v0.19.0

## Issue
When closing holepuncher it panics when it hadn't been created yet. Let me explain:

When holepuncher is closed, then the s.holePuncher is closed no matter what: 
https://github.com/libp2p/go-libp2p/blob/5151d4b4faaae04523dd83b9377881fbc717c752/p2p/protocol/holepunch/svc.go#L148-L156

It's worth noting that s.holePuncher is a pointer that is `nil` by default:
https://github.com/libp2p/go-libp2p/blob/5151d4b4faaae04523dd83b9377881fbc717c752/p2p/protocol/holepunch/svc.go#L44-L57

And that s.holePuncher is only initialized in s.watchForPublicAddr func when some conditions are met:
https://github.com/libp2p/go-libp2p/blob/5151d4b4faaae04523dd83b9377881fbc717c752/p2p/protocol/holepunch/svc.go#L141

This is a problem because if the s.holePuncher isn't initalized and we would call s.Close then s.holePuncher.Close() would panic (because it is not designed to be "zero value" closed).

## Solution:
The simplest solution is to introduce a nil check when closing holepuncher, it will check if s.holePuncher is nil or not.

### PR DESCRIPTION
**Pull request by watjurk**
This PR fixes #1448

### CODE REVIEW

**Review by marten-seemann — COMMENTED** [review]
There's another use of the `holePuncher` in `DirectConnect`. This should also be protected by the mutex.

**Comment by watjurk** [comment]
Are you sure this is what we want? In comment above direct connect we read:
```
// DirectConnect is only exposed for testing purposes.
// TODO: find a solution for this.
```
So this function should not be accesses by "normal" user

**Comment by marten-seemann** [comment]
True, but doesn’t cost us anything to lock the mutex there as well, so let’s do it.

File: p2p/protocol/holepunch/svc.go [hunk 1]
```diff
@@ -257,5 +266,7 @@ func (s *Service) handleNewStream(str network.Stream) {
 // TODO: find a solution for this.
 func (s *Service) DirectConnect(p peer.ID) error {
 	<-s.hasPublicAddrsChan
+	s.holePuncherMx.Lock()
+	defer s.holePuncherMx.Unlock()
```
Comment (marten-seemann) [inline p2p/protocol/holepunch/svc.go:270]: This locks the mutex for a **very** long time (on the order of 10s or so). Save `s.holePuncher` to a local variable instead.

**Review by marten-seemann — CHANGES_REQUESTED** [review]

File: p2p/protocol/holepunch/svc.go (hunk 1 above)
Comment (watjurk) [reply p2p/protocol/holepunch/svc.go:270]: Or maybe just use RMutex? And call RLock?

**Pull request closed without merging**